Add support for MAX_EXECUTION_TIME.

db-storage opened the pull request:
- This feature is request by [Feature 7008](https://github.com/pingcap/tidb/issues/7008).
- Add support for system variable "session.max_execution_time" and "global.max_execution_time".
- Support hint in  Select Statement linke: "SELECT /*+ MAX_EXECUTION_TIME(1000) */ * FROM t1".
- Add some unit test for the above(all tests passed).

<!--
Thank you for contributing to TiDB! Please read TiDB's [CONTRIBUTING](https://github.com/pingcap/tidb/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

### What problem does this PR solve? <!--add issue link with summary if exists-->

Fixes https://github.com/pingcap/tidb/issues/7008

### What is changed and how it works?
- If the max_execution_time is valid for  select statement(by system variable "session.max_execution_time"  or "global.max_execution_time" or hint),  then check if max_execution_time has been exceeded each time before getting new chunk, and return error if it happens.  

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

 - Unit test
> With ingested delay, in order to make query time exceeds max_execution_time.
 - Integration test

`mysql> select /*+ MAX_EXECUTION_TIME(100)*/ * from litong where SLEEP(2) and  exists (select * from test3 where test3.num>=litong.num);
ERROR 1907 (HY000): Query execution was interrupted, max_execution_time exceeded.
mysql> select * from litong where SLEEP(2);
Empty set (4.00 sec)

mysql> select * from litong;
+-----+------+
| id  | num  |
+-----+------+
| 100 |   28 |
| 101 |   28 |
+-----+------+
2 rows in set (0.00 sec)`

Code changes

 - Has interface methods change

Side effects
 
Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to be included in the release note

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/pingcap/tidb?pullRequest=10541) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/pingcap/tidb?pullRequest=10541) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/pingcap/tidb?pullRequest=10541) it.</sub>

shenli commented on server/conn.go:
```diff
@@ -1297,6 +1298,15 @@ func (cc *clientConn) writeChunks(ctx context.Context, rs ResultSet, binary bool
+		//MaxExecDuration() return 0 if there is no limit
+		if rs.MaxExecDuration().Nanoseconds() > 0 &&
+			time.Now().After(rs.StartExecTime().Add(rs.MaxExecDuration())) {
+			return errors.New("Query execution was interrupted, max_execution_time exceeded")
```
Is there an standard error code for this type of error in MySQL?

shenli replied:
If the execution is blocked at line 1311 for a long time, it may exceed the maxExecDuration.

db-storage replied:
> Is there an standard error code for this type of error in MySQL?
The original mysql error code is : "1907: Query execution was interrupted, max_execution_time exceeded.", shall the error code be  defined it in MySQLErrName(which is inside parser)?

db-storage replied:
> If the execution is blocked at line 1311 for a long time, it may exceed the maxExecDuration.
Yes, that's a problem and the current implementation is really a compromise.

I had tried closing the result set on timeout before, but that may cause a panic(due to concurrent call to Next() and Close(), no mutex protection).
If a thread is block on disk/network I/O, closing result set will not work either.
Please let me know if there is any better solution, thanks.

jackysp replied:
It is better to use this predefined error in parser repo,
https://github.com/pingcap/parser/blob/f74198ed55281727f26567edd2fae775bc6c5366/mysql/errname.go#L335

db-storage replied:
> It is better to use this predefined error in parser repo,
> 
> 
OK, I will add and use the error code. 
By the way, should all published Mysql error codes  be imported in batch?

jackysp replied:
Basically, it will be defined when it is used.

breeswish commented:
I think this should also be pushed down to TiKV to support per-statement max execution time limit.

lamxTyler commented on server/conn.go:
```diff
@@ -1297,6 +1298,15 @@ func (cc *clientConn) writeChunks(ctx context.Context, rs ResultSet, binary bool
+		//MaxExecDuration() return 0 if there is no limit
+		if rs.MaxExecDuration().Nanoseconds() > 0 &&
+			time.Now().After(rs.StartExecTime().Add(rs.MaxExecDuration())) {
+			return errors.New("Query execution was interrupted, max_execution_time exceeded")
```
But this cannot work for cases like `HashAgg` which consumes all of its child's input and finally returns result?

qw4990 replied:
It's better to start a new `goroutine` to monitor all processing queries periodically and kill these who exceed their `max_execution_time` like #10350.

qw4990 replied:
It's better to start a new goroutine to monitor all processing queries periodically and kill these who exceed their max_execution_time like #10350.

qw4990 commented on server/conn.go:
```diff
@@ -1297,6 +1298,15 @@ func (cc *clientConn) writeChunks(ctx context.Context, rs ResultSet, binary bool
+			time.Sleep(time.Duration(val.(int)) * time.Millisecond)
+		})
+		//MaxExecDuration() return 0 if there is no limit
+		if rs.MaxExecDuration().Nanoseconds() > 0 &&
```
It's better to start a new goroutine to monitor all processing queries periodically and kill these who exceed their max_execution_time like #10350.

db-storage replied:
> It's better to start a new goroutine to monitor all processing queries periodically and kill these who exceed their max_execution_time like #10350.

Added a new go routine to monitor all registered result sets. refer to the new commit.

qw4990 commented:
> I think this should also be pushed down to TiKV to support per-statement max execution time limit.

We can do this later after finishing `MaxExecTime` in TiDB.

db-storage commented:
This commit is work in progress, please ignore it. Thanks.

qw4990 commented:
@db-storage please fix CI problems and resolve conflicts.

jackysp commented:
So many unrelated changes in this PR, maybe need a squash.

db-storage commented:
> @db-storage please fix CI problems and resolve conflicts.

The last committed code depends on Error Code defined in parser, refer to [PR](https://github.com/pingcap/parser/pull/339). 

I will resolve conflicts and do a squash.

tiancaiamao commented:
This commit use context cancel to implement MAX_EXECUTION_TIME.

Cancel context is the way we used to implement `kill xxx` in the past, it has some drawbacks:
* all executor must handle the cancel signal... if some executor don't obey the rule, cancel can't work
* handle cancel signal makes the code complex
* as cancel makes the code complex... it's more likely to cause goroutine leak
* cancel will not kill 2PC (known issue)

In commit https://github.com/pingcap/tidb/pull/9844, we turn to use `executor.Close`
I think it's better to keep just one implementation, we should clean up the `cancel` way later.
Caller should use `Kill()` to terminate the executation of a session

In this PR https://github.com/pingcap/tidb/pull/10350, there is a background goroutine that monitoring all slow queries, maybe it's better to wait the PR to be merged and implement MAX_EXECUTION_TIME after that.

db-storage commented:
> This commit use context cancel to implement MAX_EXECUTION_TIME.
> 
> Cancel context is the way we used to implement `kill xxx` in the past, it has some drawbacks:
> 
> * all executor must handle the cancel signal... if some executor don't obey the rule, cancel can't work
> * handle cancel signal makes the code complex
> * as cancel makes the code complex... it's more likely to cause goroutine leak
> * cancel will not kill 2PC (known issue)
recordSet.Close() is called too when the execution is timed out.
The MAX_EXECUTION_TIME hint is permitted only for SELECT statements. Refer to [mysql doc](https://dev.mysql.com/doc/refman/5.7/en/optimizer-hints.html). So there should be no 2PC involved.

> In commit #9844, we turn to use `executor.Close`
> I think it's better to keep just one implementation, we should clean up the `cancel` way later.
> Caller should use `Kill()` to terminate the executation of a session
I can't find `executor.Close` in commit #9844，did you mean 'resultSet.Close'? 
  
> In this PR #10350, there is a background goroutine that monitoring all slow queries, maybe it's better to wait the PR to be merged and implement MAX_EXECUTION_TIME after that.
In order to use the same go routine? when will PR #10350 be merged?

tiancaiamao commented:
> I can't find executor.Close in commit #9844，did you mean 'resultSet.Close'?

Yes,  'resultSet.Close'

> In order to use the same go routine? when will PR #10350 be merged?

In order to make the code easier to maintain ... add a monitoring goroutine is a more general solution @XuHuaiyu will take charge of that PR and I guess it won't be too long get merged

shenli commented:
@db-storage Please resolve the conflicts.

db-storage commented:
> @db-storage Please resolve the conflicts.

Conflicts solved.
Rebased  code with PR #10350 merged.
Use the same handler to monitor queries with max_execution_time(by hint of system variable).
In addition to the original UT cases, added test from mysql client.

`mysql> select /+ MAX_EXECUTION_TIME(100)/ * from litong where SLEEP(2) and exists (select * from test3 where test3.num>=litong.num);
ERROR 1907 (HY000): Query execution was interrupted, max_execution_time exceeded.
mysql> select * from litong where SLEEP(2);
Empty set (4.00 sec)

mysql> select * from litong;
+-----+------+
| id | num |
+-----+------+
| 100 | 28 |
| 101 | 28 |
+-----+------+
2 rows in set (0.00 sec)`

tiancaiamao commented:
PTAL @qw4990 @XuHuaiyu

XuHuaiyu commented on util/expensivequery/expensivequery.go:
```diff
@@ -47,7 +53,8 @@ func (eqh *Handle) Run(sm util.SessionManager) {
 		select {
 		case <-ticker.C:
-			if log.GetLevel() > zapcore.WarnLevel {
+			eqh.checkAll()
```
Should we use an individual ticker here, for 2 reasons:
1. check the expensive queries every 1 second costs too much
2. maybe we need to use millisecond as the unit for max_execution_time

XuHuaiyu commented on sessionctx/variable/session.go:
```diff
@@ -379,6 +378,8 @@ type SessionVars struct {
 	// LowResolutionTSO is used for reading data with low resolution TSO which is updated once every two seconds.
 	LowResolutionTSO bool
+
+	MaxExecutionTime uint64
```
Add a comment for this attribute.

XuHuaiyu commented on session/session.go:
```diff
@@ -969,6 +966,49 @@ func (s *session) Execute(ctx context.Context, sql string) (recordSets []sqlexec
+	MaxExecutionTime = "max_execution_time"
+)
+
+type exeutionHints struct {
```
s/ exeutionHints/ executionHints

tiancaiamao commented on sessionctx/variable/tidb_vars.go:
```diff
@@ -334,7 +334,8 @@ const (
 	DefTiDBSkipIsolationLevelCheck     = false
 	DefTiDBWaitTableSplitFinish        = false
-	DefTiDBExpensiveQueryTimeThreshold = 60 // 60s
+	DefTiDBExpensiveQueryTimeThreshold = 100 // 100 ms
```
100ms is too small as the default value for `ExpensiveQueryTimeThreshold`, it makes TiDB print a log of logs.
I think 60s may be a better default value here.

And note that the check interval of the ticker is not necessary to be `ExpensiveQueryTimeThreshold / 2` because the check interval value is shared with `MAX_EXECUTION_TIME` check interval, maybe 100ms is a good choice.

It means the `MAX_EXECUTION_TIME` will be less accurate if the check interval is 100ms.  That's acceptable as the code is simple enough in my opinion. We can polish the code after this feature is merged, adopting a dynamic check interval.

tiancaiamao commented:
LGTM

PTAL @XuHuaiyu @qw4990

crazycs520 commented on sessionctx/variable/session.go:
```diff
@@ -682,6 +686,12 @@ func (s *SessionVars) SetSystemVar(name string, val string) error {
 			s.SetStatusFlag(mysql.ServerStatusInTrans, false)
 		}
+	case MaxExecutionTime:
+		timeoutMS := tidbOptInt64(val, 0)
```
How about use `tidbOptPositiveInt32`.

db-storage replied:
done

crazycs520 commented on sessionctx/variable/session.go:
```diff
@@ -692,6 +696,12 @@ func (s *SessionVars) SetSystemVar(name string, val string) error {
 		}
+	case MaxExecutionTime:
+		timeoutMS := tidbOptPositiveInt32(val, 0)
+		if timeoutMS < 0 {
```
we can remove this check now.

crazycs520 commented on sessionctx/variable/tidb_vars.go:
```diff
@@ -336,8 +336,9 @@ const (
 	DefTiDBDDLSlowOprThreshold         = 300
 	DefTiDBUseFastAnalyze              = false
 	DefTiDBSkipIsolationLevelCheck     = false
+	DefTiDBWaitTableSplitFinish        = false
```
DefTiDBWaitSplitRegionFinish       = true

crazycs520 commented on util/expensivequery/expensivequery.go:
```diff
@@ -47,38 +49,41 @@ func (eqh *Handle) SetSessionManager(sm util.SessionManager) *Handle {
 
 // Run starts a expensive query checker goroutine at the start time of the server.
-func (eqh *Handle) Run() {
+func (eqh *Handle) Run(interval time.Duration) {
```
parament of `interval ` seems useless.
and remove line#53

crazycs520 commented:
/run-all-tests

crazycs520 commented:
/run-all-tests

coderplay commented:
> > I think this should also be pushed down to TiKV to support per-statement max execution time limit.
> 
> We can do this later after finishing `MaxExecTime` in TiDB.

Did we? @zhangjinpeng1987
